### PR TITLE
Rating system

### PR DESCRIPTION
**Pull request by SameerDalal**
Created the first version of the rating system.

Currently, the user ratings are updated after every match, which is not ideal according to the glicko2 library. Instead, after a match, I want to add the result to the `matches` array and then store the array somewhere in the database. Then when the `matches` array has 10 elements, we can update the ratings and clear the array. So my question is where in the database should I store the `matches` array?

### CODE REVIEW

File: packages/server/src/ratings.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    vol: 0.06,
+  });
+  
+  const matches: [Player, Player, number][] = [];
+  
+  export async function updateRatings(
+    game: GameResponse,
+    game_obj: AbstractGame<object, object>,
+  ) {
+    // Rating system currently only for two player games
```
Comment (benjaminpjones) [inline packages/server/src/ratings.ts:22]: I notice this function has a lot of duplication (do X for player 0, do X for player 1).  It would be better to split it out into component functions like `getGlickoPlayer(id)` and `userRatingFromGlickoPlayer(player)`.  This may make it easier to unit test as well.

File: packages/server/src/ratings.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    const winner = game_obj.result[0] === "B" ? 1 : 0;
```
Comment (benjaminpjones) [inline packages/server/src/ratings.ts:23]: Is this correct?  Usually "B" is player `0`

Screenshot 2024-11-15 at 5 26 12 PM

Also, some 2-player variants (like chess) don't follow this pattern, but it's fine to ignore this for now.

File: packages/server/src/ratings.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    };
+  
+    const player_white_new_ranking: UserRanking = {
+      rating: glicko_player_white.getRating(),
+      rd: glicko_player_white.getRd(),
+      vol: glicko_player_white.getVol(),
+    };
+  
+    await updateUserRating(player_black_id, player_black_new_ranking);
+    await updateUserRating(player_white_id, player_white_new_ranking);
```
Comment (benjaminpjones) [inline packages/server/src/ratings.ts:71]: Generally when you have two async calls that don't depend on each other, it's better to execute them in parallel:

```suggestion
    await Promise.all(
      updateUserRating(player_black_id, player_black_new_ranking),
      updateUserRating(player_white_id, player_white_new_ranking),
    )
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

File: packages/server/src/games.ts [hunk 4]
```diff
@@ -199,6 +200,11 @@ export async function handleMoveAndTime(
 
   emitGame(game.id, game.players?.length ?? 0, game_obj, timeControl);
 
+  if (game_obj.phase == "gameover" && game.variant == "quantum") {
+    // user ranking value is only for quantum right now
+    await updateRatings(game, game_obj);
```
Comment (benjaminpjones) [inline packages/server/src/games.ts:205]: I'm actually not sure if we should `await` this one.  My thought process is this:

- There is no return value for us to wait for
- If there is a (rare) error in `updateRatings`, it's not such a big deal.  We have already updated the game object to reflect the move.

**Review by benjaminpjones — CHANGES_REQUESTED** [review]
Awesome!  Also could you add some information about how you are testing this?

> So my question is where in the database should I store the matches array?

I'm thinking we can store these small arrays on the user object.  Something like:

```
interface GameResult {
    opponentRating: UserRating;
    result: "win" | "loss";
}

interface GameResults {
    [variant: string]: GameResult[];
}
```

Example:

```
{
    username: "QuantumBot",
    ranking?: { rating: 1500, rd: 350, vol: 0.06 },
    recentGameResults?: { "quantum": [
      { opponentRating: { rating: 1500, rd: 350, vol: 0.06 }, result: "win" },
      { opponentRating: { rating: 1500, rd: 350, vol: 0.06 }, result: "loss" },
    ]},
}
```

But for now, I think per-game calculation is a great start!

File: packages/server/src/games.ts [hunk 5]
```diff
@@ -21,14 +21,15 @@ import {
...
+import { updateRatings } from "./ratings";
 
 export function gamesCollection() {
   return getDb().db().collection("games");
 }
 
 /**
  * @param count number of games to return (default = 10, max = 100)
- * @param offset number of games to skip (default = 0)
+ * @param offset number of games to skip (default = 0)gamesCollection
```
Comment (merowin) [inline packages/server/src/games.ts:32]: Is this addition intentional?

File: packages/server/src/games.ts (hunk 4 above)
Comment (merowin) [reply packages/server/src/games.ts:205]: Can't say I've seen many unawaited async function calls that were not mistakes. But I think it's probably not a huge deal either way, since the end of a game happens only so often.

Comment (benjaminpjones) [reply packages/server/src/games.ts:205]: That's a fair point, and I realize now the linter may complain if we don't await ([no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises/)). Let's leave as-is!

File: packages/server/src/ratings.ts (hunk 1 above)
Comment (SameerDalal) [reply packages/server/src/ratings.ts:22]: I would agree!

File: packages/server/src/ratings.ts (hunk 2 above)
Comment (SameerDalal) [reply packages/server/src/ratings.ts:23]: Yes this is correct. This `matches.push([player1, player2, winner]);` function this indicates that if winner is 1 then player1 won over player2. Since I pass player_black into the function first I am saying that player_black won/lost over player_white so "B" can be 1 or 0

File: packages/server/src/games.ts (hunk 5 above)
Comment (SameerDalal) [reply packages/server/src/games.ts:32]: It was not intentional, not sure where it came from. I'll go ahead and remove it.

File: packages/server/src/ratings.ts (hunk 2 above)
Comment (benjaminpjones) [reply packages/server/src/ratings.ts:23]: Ah, I see.  In that case, I think this variable is misleading (seems to say "Player X is the winner" when the opposite is true).  Maybe we call it `outcome`? This is what the glicko2 library names it: https://github.com/mmai/glicko2js/blob/master/glicko2.d.ts#L41


Also, should we handle ties? Quantum supports them: https://github.com/govariantsteam/govariants/blob/0abf1b20636ce4d87ed3b3e378d598a527affebb/packages/shared/src/variants/quantum.ts#L117

File: package.json [hunk 6]
```diff
@@ -24,5 +24,8 @@
   "prettier": {
     "trailingComma": "all",
     "endOfLine": "lf"
+  },
+  "dependencies": {
+    "glicko2": "^1.2.1"
```
Comment (benjaminpjones) [inline package.json:29]: This dependency should be in [`packages/server/package.json`](https://github.com/govariantsteam/govariants/blob/0abf1b20636ce4d87ed3b3e378d598a527affebb/packages/server/package.json#L12), not the top-level `package.json`.

File: packages/shared/src/api_types.ts [hunk 7]
```diff
@@ -4,10 +4,16 @@ import {
...
+  rating: number;
+  rd: number;
+  vol: number;
+}
+
 export interface User {
   username?: string;
   id: string;
-  rating?: number;
+  ranking?: UserRanking;
```
Comment (benjaminpjones) [inline packages/shared/src/api_types.ts:16]: Should we have this per variant?  Like

```
rankings: { [variant: string]: UserRanking }
```

Example:

```
{ "quantum": { rating: 1500, rd: 350, vol: rd: 0.06} }
```

File: package.json (hunk 6 above)
Comment (SameerDalal) [reply package.json:29]: The dependency was automatically added when I ran the command `yarn add glicko2`. Should I manually remove it from the top level `package.json` and add it to the `package.json` in the server folder?

File: packages/shared/src/api_types.ts (hunk 7 above)
Comment (SameerDalal) [reply packages/shared/src/api_types.ts:16]: Yes!

File: package.json (hunk 6 above)
Comment (benjaminpjones) [reply package.json:29]: > The dependency was automatically added when I ran the command yarn add glicko2.

Yeah unfortunately the "monorepo" is a bit confusing. If you are in the server/ directory when you call `yarn add` it would have added it to the correct package file.

> Should I manually remove it from the top level package.json and add it to the package.json in the server folder?

Yep, that's probably the easiest way at this point.

**Comment by SameerDalal** [comment]
All the algorithm code is completed. 
>  Also could you add some information about how you are testing this?

Currently I am using MongoDB compass to see the ranking, match history being updated. Would this be a good way of testing or is there a better way?

**Comment by merowin** [comment]
We've discussed that it should be configurable whether a game is ranked or unranked (starting with this comment:
[https://github.com/govariantsteam/govariants/issues/279#issuecomment-2212370784](url))

I just want to remind about that, and ask if we want to add this here or in a later pull request.
Steps to achieve that:
- Modify the type of the config for Quantum (perhaps introducing a new type). Ranked or not could be indicated with an optional attribute, such that it's backwards compatible with existing games.
- (Optional, depending on step 1) add the `sanitizeConfig` function for Quantum
- Add an input element for this purpose in the config form components
- Make sure the rating update code only runs for games with a "ranked config".

File: packages/server/src/games.ts [hunk 8]
```diff
@@ -199,6 +200,11 @@ export async function handleMoveAndTime(
 
   emitGame(game.id, game.players?.length ?? 0, game_obj, timeControl);
 
+  if (game_obj.phase == "gameover" && game.players.length == 2) {
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/games.ts:203]: So should I keep it as is and just exclude chess and parallel? Eg: `game.players.length == 2 && game.variant != "chess" && game.variant != "parallel"`?

File: packages/server/src/ratings.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+  game: GameResponse,
+  game_obj: AbstractGame<object, object>,
+) {
+  const variant = game.variant;
+
+  let glicko_outcome;
+  let outcome_player_black;
+  let outcome_player_white;
+
+  if (game_obj.result.endsWith("R")) {
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:32]: I see I thought the same thing. But when I go to play a game and black resigns, the result shows as 'B+R' whereas it should say 'W+R'

File: packages/server/src/ratings.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    variant,
+  );
+
+  await Promise.all([
+    updateUserRanking(player_black_id, player_black_new_ranking),
+    updateUserRanking(player_white_id, player_white_new_ranking),
+  ]);
+}
+
+async function getGlickoPlayer(
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:117]: Ah I thought that `makePlayer` returned a promise but looking at it again, the function doesn't. Thanks for catching that!

File: packages/server/src/ratings.ts [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  };
+
+  player_ranking[variant] = ranking;
+  return player_ranking;
+}
+
+function updatedGameHistory(
+  player: UserResponse,
+  opponentRating: UserRanking,
+  game_result: string,
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:150]: I see, thanks for letting me know!

File: packages/server/src/ratings.ts [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    variant,
+  );
+  const new_game_history_player_white = updatedGameHistory(
+    db_player_white,
+    db_player_black.ranking[variant],
+    outcome_player_white,
+    variant,
+  );
+
+  await Promise.all([
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:84]: I see. I thought you meant for game history we store the matches that the user has played simply for the purposes of keeping track what games a user has played. I get what you mean now.

[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:84]: Now that I think about it, do we really need a history? The third DB hit is coming from updating the history. Instead, if we pull the ranking and then update the ranking back to the DB that would take 2 hits. And since we are updating ranking after each game we wouldn't need a history.

Even if we really need to include the history we can use 1 DB hit to update both the ranking and the history at the same time.  Eg:
```
const update_result = await usersCollection().updateOne(
    { _id: new ObjectId(user_id) },
    { $set: { ranking: new_ranking, gameHistory: new_game_history } },
  );
```

[the remark this answers is not in the document]
Comment (benjaminpjones) [reply packages/server/src/ratings.ts:84]: Yeah, this was in response to your question about where to store a "matches" array:

> Currently, the user ratings are updated after every match, which is not ideal according to the glicko2 library. Instead, after a match, I want to add the result to the matches array and then store the array somewhere in the database. Then when the matches array has 10 elements, we can update the ratings and clear the array. So my question is where in the database should I store the matches array?

To be clear, this suggestion is assuming we can process one player at a time.  As described in the glicko2 docs ["How to deal with a big database of players"](https://github.com/mmai/glicko2js?tab=readme-ov-file#how-to-deal-with-a-big-database-of-players):

> ~At the end of the week~ *When we reach 10 matches for a player*, ~for each player (even those that did not play)~:
>
> - load from the database:
>     - the player *'s current* ranking data ~for the current week~
>     - the matches the player has played ~during the week~ *(from game history)*
>     - for each match, the player's opponent data ~for the current week~ *(also stored in game history)*
> - add the player and his opponents to a new glicko instance
> - add the matches to the glicko instance
> - update glicko rankings
> - save the player updated ranking data to database as the ~next week~ *new* player data

File: packages/server/src/ratings.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (benjaminpjones) [reply packages/server/src/ratings.ts:32]: You found a bug! I think it's just affecting quantum.  Let's assume no bugs here, and we can fix quantum separately (#333).

File: packages/server/src/ratings.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (benjaminpjones) [reply packages/server/src/ratings.ts:84]: just saw your new comment - no we don't need history if we update ratings one by one! Feel free to remove it and we can figure out how to improve things with a game history later.

File: packages/server/src/games.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (benjaminpjones) [reply packages/server/src/games.ts:203]: hmm,I think it should be opt-in, so new variants don't run into issues.

something like:

```
export function supportsRatings(variant) {
    return [
        "quantum",
        "baduk",
            ....
    ].includes(variant);
}
```

File: packages/server/src/ratings.ts [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+import { updateUserRanking, getUser, updateUserGameHistory } from "./users";
+
+const ranking = new Glicko2({
+  tau: 0.5,
+  rating: 1500,
+  rd: 350,
+  vol: 0.06,
+});
+
+const matches: [Player, Player, number][] = [];
```
[the remark this answers is not in the document]
Comment (benjaminpjones) [reply packages/server/src/ratings.ts:20]: > Would clearing the array (matches.length = 0) remove it from memory?

Just don't define it in the global scope

File: packages/server/src/ratings.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:32]: Awesome!

File: packages/server/src/ratings.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/ratings.ts:84]: Got it!

**Comment by benjaminpjones** [comment]
> > Also could you add some information about how you are testing this?
> Currently I am using MongoDB compass to see the ranking, match history being updated. Would this be a good way of testing or is there a better way?

That's great, as long as there is some way to know it's working as intended

> is there a better way?

Of course, some automated (unit) tests would be great too - let me know if you'd like to try adding them and if any guidance is needed.

Some examples of tests in server/ code: https://github.com/govariantsteam/govariants/tree/26506e38cfcf1a17f16cdc4eb6abeb37ec22c5b8/packages/server/src/time-control/__tests__

**Comment by SameerDalal** [comment]
Will do!

**Comment by SameerDalal** [comment]
I added the unit test cases for the functions.

Just wanted to clarify something about the unit tests - the unit tests will only really be valuable when the code in `ratings.ts` is changed right? Right now, the tests are only checking if the functions are working as intended, but they don't give an indication of the values being updated in the database.

Also Glicko has its own unit tests which we can run using the command `npm run test`. But when running the command, I don't see the "PASS" message for the Glicko tests specifically.

**Review by benjaminpjones — APPROVED** [review]
Hey, thanks for addressing all the feedback - this will be a great feature to have!  BTW I am introducing "user pages" in #345, so that's at least one place we can display the rating in the future.

Left a couple comments on the new tests, but I think they look pretty good already :)

File: packages/server/src/rating/__tests__/rating.test.ts [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+  const db_player_ranking_quantum = db_player.ranking["quantum"];
+
+  const ranking = new Glicko2({
+    tau: 0.5,
+    rating: 1500,
+    rd: 350,
+    vol: 0.06,
+  });
+
+  const new_ranking = new Glicko2({
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/rating/__tests__/rating.test.ts:51]: I believe we do. The Glicko object has some internal variables like `id` that change when we use the object. Since the `.toEqual` function is doing a deep equality, it's comparing all the values in the object including `id`. So, if we didn't have the new rating object the test would fail even though we only want to compare the rating, rd, and vol. Instead we could use the function `toMatchObject` instead of `toEqual` and only compare the properties that we need to compare.

File: packages/server/src/rating/__tests__/rating.test.ts [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    tau: 0.5,
+    rating: 1500,
+    rd: 350,
+    vol: 0.06,
+  });
+
+  expect(getGlickoPlayer(db_player_ranking_quantum, ranking)).toEqual(
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/rating/__tests__/rating.test.ts:58]: Got it!

File: packages/server/src/rating/__tests__/rating.test.ts [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+  const new_quantum_ranking: UserRanking = {
+    rating: 1667,
+    rd: 295,
+    vol: 0.0699,
+  };
+
+  const glicko_player = getGlickoPlayer(new_quantum_ranking, glicko_ranking);
+
+  const expected_player: UserResponse = {
```
[the remark this answers is not in the document]
Comment (SameerDalal) [reply packages/server/src/rating/__tests__/rating.test.ts:135]: You're right, we don't need the whole player object!
